### ISSUE TITLE
Social connections disabled if enabled_connections property omitted

### ISSUE THREAD

**Issue by wp55** [auth0/auth0-deploy-cli#523]
In the past we've always used AUTH0_EXCLUDED_CLIENTS because we don't want to script the applications. So we've added all the applications in the excluded clients list in this way:

 "AUTH0_EXCLUDED_CLIENTS": [
    "application 1",
    "application 2",
    "..."
  ]

As this is being deprecated in the future, I've started implementing the new AUTH0_EXCLUDED in this way:

"AUTH0_EXCLUDED": [
    "clients",
    "clientGrants",
    "resource_servers"
  ]

Now we see that after deploy all connections on all applications are dropped. Something which was never the case when we used AUTH0_EXCLUDED_CLIENTS.

Is it possible to have a fix for this?

Thank you

**Comment by willvedd** [comment]
Sorry to hear that your connections dropped and appreciate you opening the ticket. Am I correct in understanding that you've excluded connections via the `AUTH0_EXCLUDED` property, something you previously did via the `AUTH0_EXCLUDED_CLIENTS` property, and you've found that _connections_ have been deleted as a result? If so, that's odd and I don't believe there to be a correlation.

Firstly, can you confirm that you're using v7.7.0 or later? I'm assuming that if you're seeing deprecation warnings that `AUTH0_EXCLUDED` is supported, but just want to make sure.

Secondly, can you provide the sample configuration for both clients and connections? My hunch is that perhaps you've set connections to an empty array which is interpreted as intentionally setting as empty.

**Comment by wp55** [comment]
Yes, we're using v7.7.0

No, we don't exclude connections. In AUTH0_EXCLUDED we've only included applications (clients), clientGrants and resource_servers (apis). As we want to manually deploy these 3 types. So it looks like this:

"AUTH0_EXCLUDED": [
"clients",
"clientGrants",
"resource_servers"
]

In our tenant.yaml the definition of the connections is included and also deployed this way:

![image](https://user-images.githubusercontent.com/87134268/166629593-5e56055f-cd13-4e0a-8c22-b5a69591de0d.png)

When we manually configure an application, we also manually configure the connections that are being used with that application. After deploy with the new setup (AUTH0_EXCLUDED is used instead of the old AUTH0_EXCLUDED_CLIENTS) the connections have dropped from the applications. In the past when we used AUTH0_EXCLUDED_CLIENTS these connections were never dropped from the applications.

Regards,
Wim

**Comment by willvedd** [comment]
I'd like to understand why this issue is happening because it surely sounds nefarious, but I'm unable to reproduce this issue. Here is what I'm doing:

1. Creating social connections manually in tenant
2. Export/dump configurations with clients and client grants excluded via `AUTH0_EXCLUDED` (`resource_servers` doesn't do anything because of the underscore)
3. Import the configurations exported from step 2, with clients and client grants excluded via `AUTH0_EXCLUDED`.

Despite the above, which seems to match your description, the connections remain. In fact looking at the code, it appears that the only overlap between connections and clients is to determine if a connection has clients attached for determining if it's enabled or not.

As I mentioned above, when you're passing `resource_servers` into `AUTH0_EXCLUDED`, it won't work because of the underscore, if you wish to exclude resource servers too, you'll need to list it as `"resourceServers"`. I'm not sure if that's related to your issues, but that's an easy thing to fix right away.

Because I can't recreate this issue, can you provide step-by-step instructions I can follow?

**Comment by willvedd** [comment]
Ah ok, thanks for detailing that for me. I misunderstood what you meant by "drop", in your case it means that the connections are getting disabled. I'll be able to look at this in the coming days.

**Comment by willvedd** [comment]
Had a chance to look into this. Don't think it's a bug with `AUTH0_EXCLUDED`. From what I can tell, connections and clients do have a relationship but excluding one will not have an impact on the other. In your above example, you omitted the `enabled_clients` property from the database connections. The existence of enabled clients on a connection is what determines the state of that UI toggle switch. 

As a proof of concept, I imported the above sample configuration and then exported back. The `enabled_clients` property was provided. As a test, I set one as an empty array:

```yaml
connections:
  - name: facebook
    strategy: facebook
    enabled_clients:
      - ce2p3bRuQ***********fxdJYaSuJb
    is_domain_connection: false
    options:
      email: true
      scope: email
      client_id: xxxxxxxxxxxxxxxxxx
      client_secret: xxxxxxxxxxxxxxxxxxxxxxx
  - name: google-oauth2
    strategy: google-oauth2
    enabled_clients: []
    is_domain_connection: false
    options:
      email: true
      scope:
        - email
        - profile
      profile: true
      client_id: xxxxxxxxxxxxxxxxxxxxxxxxxxx
      client_secret: xxxxxxxxxxxxxxxxxxxxxxx
  - name: linkedin
    strategy: linkedin
    enabled_clients:
      - ce2p3bRuQ***********fxdJYaSuJb
    is_domain_connection: false
    options:
      email: true
      scope:
        - r_emailaddress
        - r_liteprofile
      profile: true
      client_id: xxxxxxxxxxxx
      client_secret: xxxxxxxxxxx
      strategy_version: 2
```

Note the `google-oauth-2` has an empty array set for enabled clients. Applying the configuration resulted in:

![Screen Shot 2022-05-25 at 8 42 35 AM](https://user-images.githubusercontent.com/4614315/170264577-04457443-cc1d-4602-a586-1612d0486ea5.png)

This behavior is confirmed in the [documentation for updating connections](https://auth0.com/docs/api/management/v2#!/Connections/patch_connections_by_id):

> The identifiers of the clients for which the connection is to be enabled. If the property is not specified, no clients are enabled. If the array is empty, the connection will be disabled for every client.

To conclude, I don't think this is a bug caused by `AUTH0_EXCLUDED`, but rather the `enabled_clients` property on your connections not being properly set.

**Comment by wp55** [comment]
You are referring to a patch operation. A patch operation which has an optional field "enabled_clients". Which means that if enabled_clients is not present, nothing is done with it. That's the definition of a patch. We don't include an empty array for enabled_clients so nothing should be done with it and the enabled clients should stay as-is. So they should stay enabled.

You are introducing AUTH0_EXCLUDED because then we don't have to include applications anymore in AUTH0_EXCLUDED_CLIENTS. But now you're saying that we have to include all those client_id's in scripting. This means I will have to add them as variables into AUTH0_KEYWORD_REPLACE_MAPPINGS. Because we have 4 environments per auth0 tenant (dev, test, uat and prd) and 1 deploy script per tenant. This is worse than before when we had to work with AUTH0_EXCLUDED_CLIENTS.

**Comment by willvedd** [comment]
> You are referring to a patch operation. A patch operation which has an optional field "enabled_clients". Which means that if enabled_clients is not present, nothing is done with it.

This is a fair point. Looking at the code I see that undefined values are coerced into an empty array, which would cause the above symptoms. And as far as I can tell, it's the only cause for your issues, the resource exclusion mechanism is completely unrelated. 

I think it's a reasonable expectation to omit `enabled_clients` from the API payload if is omitted from the resource configuration files. Fix should be relatively straightforward; will hopefully get around to this in the near future.

**Comment by willvedd** [comment]
The fix for this just got released in [v7.13.0](https://github.com/auth0/auth0-deploy-cli/releases/tag/v7.13.0). I recommend updating and giving it a go. Appreciate the bug ticket!

**Comment by wp55** [comment]
This seems to work as expected now!
Thanks
Wim
